cli/azd/pkg/ux/canvas.go

```diff
@@ -0,0 +1,94 @@
+package ux
+
+import (
+	"io"
+	"sync"
+)
+
+// Canvas is a base component for UX components that require a canvas for rendering.
+type canvas struct {
+	visuals    []Visual
+	printer    Printer
+	writer     io.Writer
+	renderMap  map[Visual]*VisualContext
+	updateLock sync.Mutex
+}
+
+type Canvas interface {
+	Run() error
+	Update() error
+	WithWriter(writer io.Writer) Canvas
+}
+
+// NewCanvas creates a new Canvas instance.
+func NewCanvas(visuals ...Visual) Canvas {
+	canvas := &canvas{
+		visuals:   visuals,
+		renderMap: make(map[Visual]*VisualContext),
+	}
+
+	return canvas
+}
+
+// WithWriter sets the writer for the canvas.
+func (c *canvas) WithWriter(writer io.Writer) Canvas {
```

Comment:
Currently `WithWriter` is used to set `c.writer`, but `c.writer` is only used when `Run()` is called. The effect is that calling `WithWriter` after `Update()` doesn't actually change any behavior -- 

does it make more sense to:
1.  Have `Writer` be part of the constructor, or
2. Document this ineffectual behavior?

Reply:
The canvas component and the `Run` function are used internally by each UX component.  
If the component is not rendering inside a larger multi-component canvas it creates its own internal canvas for the main rendering loop out to the default writer.